Enhancement Proposal for Adding HTTP header to HAProxy Without Customizing haproxy.config Template (#1296)
openshift/enhancements: enhancements/ingress/set-delete-http-headers.md

this gets decoded into the two triples ("Accept", "text/plain, text/html", "Set") and ("Accept-Encoding", "Delete"); then the template adds the following stanzas to the `public`, `fe_no_sni`, and `fe_sni` frontends:
```conf
    http-response set-header X-Frame-Options 'DENY'
    http-response set-header X-XSS-Protection '1;mode=block'
    http-response set-header x-forwarded-client-cert '%{+Q}[ssl_c_der,base64]'
    http-response del-header X-Frame-Options
    http-response del-header X-XSS-Protection
    http-request set-header Accept 'text/plain, text/html'
    http-request del-header Accept-Encoding
```
Header values are quoted so that HAProxy does not misinterpret or throw an error while parsing `http-response` and `http-request` stanzas in `haproxy.config`
that contain spaces or other special characters.

> Any single quotes provided within the header value will be escaped and replaced with `'\''`, and then the value is encapsulated with single quotes `''`.

Author: '' is correct.

Header value `%[req.hdr(Host)] if foo` is quoted as `'%[req.hdr(Host)] if foo'`.

The following shows how it will get set in haproxy.config
```conf

> http-response set-header X-Frame-Options '"abc"\ '\''def'\'''

Reviewer: nit: There is no mention of the first example in the previous section.

    // If this field is empty, no action is taken.
    // +optional
    // +unionMember
    Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
}
```

```go
// IngressControllerHTTPHeaderActionType defines actions that can be performed on HTTP headers.
type IngressControllerHTTPHeaderActionType string

const (
	// Set specifies that an HTTP header should be set.
	Set IngressControllerHTTPHeaderActionType = "Set"
	// Delete specifies that an HTTP header should be deleted.
	Delete IngressControllerHTTPHeaderActionType = "Delete"
)
```

The `IngressControllerSetHTTPHeader` type has a field for specifying a header value.
```go
// IngressControllerSetHTTPHeader defines the value which needs to be set on an HTTP header.
type IngressControllerSetHTTPHeader struct {
	// value specifies a header value.
	// Dynamic values can be added. The value will be interpreted as an HAProxy format string as defined in
	// https://cbonte.github.io/haproxy-dconv/2.4/configuration.html#8.2.4 and may use HAProxy's %[] syntax and
	// otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.

> 	// It must be no more than 32768 characters in length.
> 	// This value was selected as most of the servers accept this value.

Reviewer 2: Please add a note that the total size of all added headers _after_ interpolating dynamic values must not exceed the rewrite buffer size.  Please add the note in these four places for visibility:
* The `value` field of `IngressControllerSetHTTPHeader` (i.e. here).
* The `value` field of `RouteSetHTTPHeader`.
* The `actions` field of `IngressControllerHTTPHeaders`.
* The `actions` field of `RouteHTTPHeaders`.
```suggestion
	// The value of this field must be no more than 32768 characters in length.
	// This limit was selected as most common web servers have a limit of 16384 characters
	// or some lower limit.
	//
	// Note that the total size of all added headers *after* interpolating dynamic values
	// must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
	// IngressController.
```

the `haproxy.config` sections corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ([code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).

The router sets this variable based on the Route's `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` field values.
`ServiceAliasConfig` is populated by [the router's `createServiceAliasConfig` method](https://github.com/openshift/router/blob/3065f6583f3925328fbdbfe95e3bc7bb7a084d33/pkg/router/template/router.go#L908-L935)

```go
// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
type ServiceAliasConfig struct {
    ...
    // For route based setting of http response headers
    HTTPResponseHeaders []HTTPHeader

    // For route based setting of http request headers
    HTTPRequestHeaders []HTTPHeader
}
```
```go
// HTTPHeader specifies an HTTP header that should be set/replace/append

> type HTTPHeader struct {
> // Name specifies an HTTP header name.
> Name string
> 
> // Value specifies the header value.
> Value string
> 
> // Action specifies the action to be performed.
> Action string
> }

Reviewer: nit
```suggestion
type HTTPHeader struct {
   // Name specifies an HTTP header name.
   Name string

   // Value specifies the header value.
   Value string

   // Action specifies the action to be performed.
   Action string
}
```

Author: done

```

> In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside the router. So, we will run a loop
> on either `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type` or `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type` which will have the desired
> sequence. 
> 
> If request and response headers are provided in the API then first we run loop through request headers and then response headers.

Reviewer: I think this says the same thing, a little more clearly.
```suggestion
The sequence of action operations will be maintained according to the router RouteSpec. First, we iterate  over `spec.httpHeaders.actions.request[*].action.set` or `spec.httpHeaders.actions.request[*].action.type`, then over `spec.httpHeaders.actions.response[*].action.set` or `spec.httpHeaders.actions.response[*].action.type`.
```

Author: done

C) library-go code will be having validations for checking Route API fields for OCP.
   We will be creating an immutable/hard-coded deny-list of HTTP response headers in [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141)
   which won't be allowed to be set or deleted. 

>    For the initial enhancement, the HSTS `strict-transport-security` header name is included in this deny-list.

Reviewer: ```suggestion
   For the initial enhancement, the HSTS `strict-transport-security` and the `proxy`  header name are included in this deny-list.
```

Author: I added it in a different way.

   
   Only 64 items in request and response header list will be allowed.
   
   Minimum length of the header value would have to be greater than 0 and less than 4096.
   
   Sample fetchers for response dynamic header values allowed are "res.hdr" and "ssl_c_der".
   
   Sample fetchers for request dynamic header values allowed are "req.hdr" and "ssl_c_der".
   
   Converters allowed for request and response are "lower" and "base64".
   

>   Actions like Set/Delete for headers can't be set for passthrough routes.

Reviewer: ```suggestion
  Actions like Set/Delete for headers are ignored for passthrough routes.
```

Reviewer 2: I believe "can't be set" is more accurate—validation catches attempts to specify custom headers for passthrough routes:
```go
    // spec is the desired state of the route
    // +kubebuilder:validation:XValidation:rule="!has(self.tls) || self.tls.termination != 'passthrough' || !has(self.httpHeaders)",message="header actions are not permitted when tls termination is passthrough."
    Spec RouteSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
```
If you want this to be more explicit, how about this?
```suggestion
  Validation prohibits configuring any header actions on passthrough routes.
```

Reviewer: Does `||` mean something other than `or` here?

Author: it is a rule. condition or

Author: i still think `Actions like Set/Delete for headers can't be set for passthrough routes.` is more better

Author: Validation prohibits configuring any header actions on passthrough routes.  added this as looks more better

D) route/v1 of openshift/api will specify kubebuilder tags and CEL to enable client-side validation.

E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be the same as mentioned in point C) except the one for passthrough routes.

F) To validate the header value it has to match a regex.
For the IngressController CRD and Route CRD, this validation is defined on the request and response action lists in openshift/api.
For the Route API as handled by openshift-apiserver, this is defined in library-go.
This will be the regex:
```go
For request header:
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
```go
For response header:
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```

> Allowed dynamic values for request would be -

Reviewer: ```suggestion
Example dynamic values for request would be -
```

> Allowed dynamic values for request would be -

Reviewer: ```suggestion
G) Dynamic values are allowed.  (more details can be found in .....)?
Example dynamic values for request:
```

```conf
%[req.hdr(host),lower]
%[req.hdr(host),base64]
%[req.hdr(Host)] if foo
```

> Allowed dynamic values for response would be -

Reviewer: ```suggestion
Example dynamic values for response would be -
```

Author: done

```conf
%[res.hdr(X-Acl-Key),lower]
%[res.hdr(X-Acl-Key),base64]
%[req.hdr(Host)] if foo
```

> Dynamic values allowed for both request and response would be -

Reviewer: ```suggestion
Example dynamic values that can be used by both request and response:
```

```conf
%{+Q}[ssl_c_der,base64]
```

> For regular header which are not dynamic one can pass values as mentioned in unit test `TestHeaderValues` with `validInput:true`

Reviewer: ```suggestion
H) Static header values
For regular headers which are not dynamic one can pass values as listed in the unit test `TestHeaderValues` with `validInput:true`
```

Author: done

present in `pkg/route/validation/validation_test.go`

> Please note that all values will be quoted before interpolating them into `haproxy.conf`.  Specifically, any single-quote present in the value is escaped, and then the whole value
> is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. This quoting logic will be present in the router.
> To quote using single-quotes, we will do the following:
> 1. Replace `'` with `'\''`.
> 2. Wrap the entire string in single quotes.
> For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.

Reviewer 2: I see Candace's point—someone could read the current text ("all values will be quoted") and believe that it means that the _user_ will/must quote all values.  How about this?
```suggestion
Please note that the router quotes all values before interpolating them into `haproxy.conf`.  The user must not add quotes in the API `value` field unless the user wants literal quotes in the HTTP header value.  Specifically, any single-quote present in the value is escaped, and then the whole value
is surrounded in single-quotes.  This prevents HAProxy config parsing errors in the event of invalid input. This quoting logic will be present in the router.
To quote using single-quotes, we will do the following:
1. Replace `'` with `'\''`.
2. Wrap the entire string in single quotes.
For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.
```

> For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.

Author: yeah this is good if we specifically want to tell users about quotes in values.

> For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.

Author: done

### Risks and Mitigations

We identified the following headers that require special explanations:

#### Proxy header
The `proxy` HTTP request header can be used to exploit vulnerable CGI applications using the [_httpoxy_](https://httpoxy.org/) vulnerability.
In this vulnerability, if the client sets the HTTP `proxy` request header and the proxy passes it through to the application, then the CGI application takes the header value and injects it in the `HTTP_PROXY` environment variable.

In this manner, an attacker could direct a vulnerable application to use a proxy of the attacker's choice.
Note that OpenShift router already mitigates this vulnerability by deleting any `proxy` header that the client specifies.

> Allowing the cluster-admin or route owner to override the `proxy` HTTP header isn't necessarily a security risk; it would only allow admins or route owners to inject configuration into their own applications, which they already control.  However, the `proxy` header is non-standard, and allowing it to be configured is potentially
> confusing.  For this reason, we do not allow setting the `proxy` header with this EP.

Reviewer: ```suggestion
Allowing the cluster-admin or route owner to override the `proxy` HTTP header isn't necessarily a security risk; it would only allow admins or route owners to inject configuration into their own applications, which they already control.  However, the `proxy` header is non-standard, and allowing it to be configured is error-prone.  For this reason, we have not and still do not allow setting the `proxy` header.
```

Author: i think confusing word is more better than error-prone.  I think the original text is fine.

Author: i think your suggestion looks better. So done.

#### Strict-Transport-Security

> Strict-transport-security is already handled separately, and we don't want to
> allow overrides of that configuration.

Reviewer: ```suggestion
Strict-transport-security is already [handled separately](https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-enabling-hsts_route-configuration), so we don't allow overrides of that configuration.
```

If we allowed it, then route owners could impact each other's routes.

> It must be blocked so, we have put this
> header in the deny list.

Reviewer: ```suggestion
It must be blocked so, we do not allow setting or deleting the `strict-transport-security` header in the ingress specification.  It is still allowed on the route.
```

> Strict-transport-security is already handled separately, and we don't want to
> allow overrides of that configuration.
> 
> If we allowed it, then route owners could impact each other's routes.
> 
> It must be blocked so, we have put this
> header in the deny list.

Reviewer: ```suggestion
Strict-transport-security is already [handled separately](https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-enabling-hsts_route-configuration), so we don't allow overrides of that configuration.  We don't allow setting the `strict-transport-security` headers on either ingress or route.
```

Author: done

#### X-Forwarded-For
`x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
and possibly a warning on admission.

For example: If someone creates a Ingress Controller with the following:
```sh
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpHeaders:
    forwardedHeaderPolicy: Replace
```
Then the `set-header` stanzas will be added in `haproxy.config` using the [haproxy-config.template](https://github.com/openshift/router/blob/0cd9a58dff77127562f7851ccfa5cda79b8cef48/images/router/haproxy/conf/haproxy-config.template#L565-L580)
However, if someone wants to set custom values for these then they need to use the API of this EP.

> If both are set then custom values will take over.  

Reviewer: There is also an annotation on the Route:  `haproxy.router.openshift.io/set-forwarded-headers`

Can you please list the four different ways `x-forwarded-for` can be specified, and how these ways interact in processing?

Reviewer: A table would be excellent.  You could list operations too, to specify that deleting is only allowed by this new feature.

Reviewer: We talked about this in the setheaders slack channel.

Author: done

Please note that the setting of forwardedHeaderPolicy happens at the plain backend section of haproxy.config i.e. per route level and not at the front-end section
of the haproxy.config level as per the above-mentioned link.

Another use case is deleting the header which forwardedHeaderPolicy does not provide but this EP does.

#### X-SSL
Similarly to `x-forwarded-for`, there are legitimate use-cases for setting `x-ssl` and related headers, and allowing the cluster-admin or route owner to modify these headers presents no security risk.
For example: HAProxy takes information from TLS handshake, tls session or the certificate that client provided. These headers are used for mtls and client certificates.

> These headers tell you whether the client provided a certificate, was the certificate valid, what was the content or cipher, distinguished name or expiration date
> on the certificate using converters with prefix like `ssl_` which can be used in the haproxy proxy config to take that information and put it http request headers which get sent

Reviewer: I thought there were no other converters than the ones listed.  What converters are you talking about here?
```suggestion
These headers tell you whether the client provided a valid certificate, and certificate values such as content,  cipher, distinguished name, or expiration date on the certificate.  

```

Author: The original text looks good to me. or is used correctly from my understanding.  It is just to explain why ssl related fetchers can't be an issue.

Author: i have made some changes.

Reviewer 2: > I thought there were no other converters than the ones listed. What converters are you talking about here?

HAProxy has scores of converters, documented here: <http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#7.3.1>.  The reason why this enhancement doesn't allow more converters is explained in "Should this enhancement allow setting dynamic header values?" under "Open Questions".

to the backend application.

You must already trust the cluster-admin or route owner, who has full control over the proxy or application.  Thus, there is no additional risk in trusting
an `x-ssl` header that is controlled by the cluster-admin or route owner.  For this reason, we do not restrict setting this header.

#### Host

> Setting the `host` header via the IngressController CR can cause HAProxy to fail to look up the correct route, and for that reason, we prohibit setting the `host` request header on the IngressController.

Reviewer 2: Custom headers that are specified on the route get configured on the corresponding backend, and the backend configuration takes effect after the frontend has done its map lookup.

> Setting the `host` header via the IngressController CR can cause HAProxy to fail to look up the correct route, and for that reason, we prohibit setting the `host` request header on the IngressController.

Author: Added this in open questions

OpenShift router's `haproxy.config` file uses [the `base` sample fetch method](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#7.3.6-base) in the frontends to get the
HTTP request `host` header and path, which are then used with [the `map` converter](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#7.3.1-map) to look up the
appropriate backend to use for the request.

Modifying the `host` header on the IngressController therefore could cause HTTP requests to fail with HTTP 503 errors.
For this reason, the `host` header is in the deny list of headers that are prohibited to be set or deleted using the IngressController API.

> ### Drawbacks

Reviewer: I would list the potential difficulty in documenting and troubleshooting this as a drawback, e.g. all the special treatment of certain headers.

Author: i have pointed this section to `Risks and Mitigation` section

## Design Details

### Open Questions

#### Will this enhancement cover setting or deleting request headers in addition to response headers?
  Answer: Yes, we will be covering setting or deleting of request headers.
  This is required for the following use-case.  
  
  As a cluster admin or route owner, I want to enable mutual TLS but replace the "x-ssl-client-der" request header
  that OpenShift router adds with an "x-forwarded-client-cert" request header that my application checks for, 
  so that I can migrate my applications that rely on mutual TLS and use the latter header name. 

  The following example replaces "x-ssl-client-der" with "x-forwarded-client-cert":
```yaml
spec:
  httpHeaders:
    actions:
      response:
      - name: X-SSL-Client-Cert
        action:
        type: Set
        set:
          value: "%{+Q}[ssl_c_der,base64]"
      - name:   Content-Language
        action:
        type: Delete
```

#### Will this enhancement cover setting the load-balancing algorithm based on headers?

>   Answer:The HAProxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the

Reviewer: nit
```suggestion
  Answer: The HAProxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the
```

Author: done

Author: done

  check against the incoming request's header values.
  
  Even though this enhancement adds the header manipulation capability it
  doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route. 
  
  Therefore, such a load balancing algorithm is out of scope for this enhancement and should be handled with a separate RFE and thereby it will be considered as a non-goal.

#### Will this enhancement cover deleting or appending headers?
  Answer: Yes and no.  This enhancement will cover deletion but not appending of
  HTTP response headers as per the requirements from
  [RFE-464](https://issues.redhat.com/browse/RFE-464). 
  While this RFE mentions use-cases that require the ability to delete headers, there is no use-case
  from this RFE that requires the ability to append headers. 

  Once such a use-case arises, we may add the option to append headers, with due
  consideration of any additional requirements deriving from the use-case
  itself or considerations around the handling of comma-separated values.  (Note
  that a use-case involving the `x-forwarded-*` headers, which is not mentioned
  on the RFE, is described below.)
  

>   "set-header" does the same as "add-header" except that the header name is first removed if it existed. This is 
>   useful when passing security information to the server, where the header must not be manipulated by external users.
>   So, that is the reason we don't need an add-header.
>   References - https://wohugb.gitbooks.io/haproxy/content/config/http-response.html
>        https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/

Reviewer: ```suggestion
  "set-header" can be used for the same purpose as "append-header" except that the user must know the value before adding to it. This is 
  useful when passing security information to the server, where the header must not be manipulated by external users.  We chose to use "set-header" instead of "append-header" to provide a measure of security.
  References - https://wohugb.gitbooks.io/haproxy/content/config/http-response.html
       https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/
```

Reviewer 2: The original statement is more accurate: `set-header` tells HAProxy to overwrite any existing header; that is, remove the header if it is already set and then add the header with the custom value.  The user does not need to know the old value to overwrite it with the custom value.  

By the way, is the comment about security important?  It seems confusing to me; if the point about security is really important, it should be clarified, and otherwise it can be omitted.

Author: ok removed it. that sounds right.

Reviewer: We already use add-header in https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L554 and lines after, dealing with x-forward and forward values.  This section doesn't make sense. The question should just be  `Will this enhancement cover appending headers` and no mention of deleting, which is already covered.

  
  For the headers like `X-Forwarded-For` which might need values to be appended we have a different API for that.

>   This API is specifically meant to just either set or delete an HTTP Request and Response header.

Reviewer: Move this to the next section if needed there.

Author: ok

  In this API if someone sets an HTTP header with a value we expect them to set a complete value in advance rather than appending it later to the header.

>   Also using this API one can specify the header name in the list only once to avoid providing two entries of the same header name.

Reviewer: Not sure this line belongs here.

Author: it belongs here as I need to explain that we don't allow append so what one needs to know while setting a header.
Also while setting what is the uniqueness validation.

#### For headers such as `X-Forwarded-For`, it makes sense to add an "Append" action. So why does this enhancement not cover appending?
  Answer: Currently, we are proceeding with Set and Delete only. The RFE requires these two only. We have confirmed this with the stakeholders as well.

>   For the `X-Forwarded-*` headers, we have a dedicated API field for IngressControllers and a Route annotation, which provide the equivalent of an `Append` action for these headers:

Reviewer 2: There is a recognized "append" value for the annotation:
https://github.com/openshift/router/blob/6cef4d006b96779a78d752909203812a451e8d27/images/router/haproxy/conf/haproxy-config.template#L491-L494
https://github.com/openshift/router/blob/6cef4d006b96779a78d752909203812a451e8d27/images/router/haproxy/conf/haproxy-config.template#L551-L568

```sh
$ oc explain ingresscontroller.spec.httpHeaders.forwardedHeaderPolicy
KIND:     IngressController
VERSION:  operator.openshift.io/v1

FIELD:    forwardedHeaderPolicy <string>

    DESCRIPTION:
    forwardedHeaderPolicy specifies when and how the IngressController sets the
    Forwarded, X-Forwarded-For, X-Forwarded-Host, X-Forwarded-Port,
    X-Forwarded-Proto, and X-Forwarded-Proto-Version HTTP headers. The value
    may be one of the following:
    * "Append", which specifies that the IngressController appends the headers,
    preserving existing headers.
    * "Replace", which specifies that the IngressController sets the headers,
    replacing any existing Forwarded or X-Forwarded-* headers.
    * "IfNone", which specifies that the IngressController sets the headers if
    they are not already set.
    * "Never", which specifies that the IngressController never sets the
    headers, preserving any existing headers. By default, the policy is
    "Append".
```

#### Are some headers blocked from being set or deleted?

>   Answer: Please refer to the `Risks and Mitigations` section for more details.

Reviewer: ```suggestion
  Answer: Yes, please refer to the `Risks and Mitigations` section for more details.
```

Author: done

> #### Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). They usually want to cache the static assets on the client and do not care about session stickiness for those requests. Only the dynamic requests need to stick to the same server (session sticky)
> Most applications will set something like this for static assets:
> ```sh
> cache-control: max-age=7200, must-revalidate, public
> ```
> But the OpenShift router (HAProxy) will add cache-control: private which results in both headers in the response:
> ```sh
> cache-control: private
> cache-control: max-age=7200, must-revalidate, public
> ```
> This leads to issues with caching: CDNs like Cloudflare will not cache your files.
> Can we please make sure to only add the `cache-control: private` header if there is no `cache-control: public` header already present in the response?
>   Answer: A common technique is to put static assets on a different virtual host. In OpenShift terms, that means using a distinct route for static assets. This route can
>   be annotated with `haproxy.router.openshift.io/disable_cookies=true` to prevent setting the cookie and thus prevent setting cache-control.

Reviewer: I think this is what you want to say here, but not sure.  It doesn't really explain why the answer is "no".

```suggestion
#### Can we please make sure to only add the `cache-control: private` header if there is no `cache-control: public` header already present in the response?

  Answer:  No.  Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). They usually want to cache the static assets on the client and do not care about session stickiness for those requests. Only the dynamic requests need to stick to the same server (session sticky).

Most applications will set something like this for static assets:
```sh
cache-control: max-age=7200, must-revalidate, public
```
But the OpenShift router (HAProxy) will add cache-control: private which results in both headers in the response:
```sh
cache-control: private
cache-control: max-age=7200, must-revalidate, public
```
This can lead to issues with caching: e.g. CDNs like Cloudflare will not cache your files.

  As a workaround, a common technique is to put static assets on a different virtual host. In OpenShift terms, that means using a distinct route for static assets. This route can
  be annotated with `haproxy.router.openshift.io/disable_cookies=true` to prevent setting the cookie and thus prevent setting cache-control.
```

Author: done

#### Would we be able to [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression) or [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression)?
  Answer: No, this is out of scope for this enhancement. Action `Set` will perform `set-header` in the haproxy.config for `http-response` and `http-request` which will either set the value if the value is not present or replace the existing value.

> #### Should this enhancement allow setting dynamic header values?
>   Answer: Yes, you would be able to set a minimal set of dynamic header values. 
>   We had three options for this EP.
>   1. Allow dynamic header values with all sample fetchers and converters.
>   2. Allow a minimal set of dynamic values that we can confidently validate, focusing on the things that people have asked for on the RFE or [EP](https://github.com/openshift/enhancements/pull/1296#issuecomment-1396103783).
>   3. Prohibit dynamic values entirely.

Reviewer: ```suggestion
#### Should this enhancement allow setting dynamic header values?
  Answer: Yes, you would be able to set a minimal set of dynamic header values. 
  We had three options for this EP.
  
  1. Allow dynamic header values with all sample fetchers and converters.
  2. Allow a minimal set of dynamic values that we can confidently validate, focusing on the things that people have asked for on the RFE or [EP](https://github.com/openshift/enhancements/pull/1296#issuecomment-1396103783).
  3. Prohibit dynamic values entirely.
```

Author: do we really need a new line ?

Author: anyway done

Reviewer: Adding the blank line makes the list show up as an ordered list. Without it, they are all within the previous paragraph.

  However, as few customers asked for it, we went ahead with option number 2.

  For setting dynamic header values for HTTP responses, allowed sample fetchers are `res.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
  
  For setting dynamic header values for HTTP requests, allowed sample fetchers are `req.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
    
  Example: One might be able to set the value of `x-forwarded-client-cert` to `%{+Q}[ssl_c_der,base64]`.

> #### Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation along with a setting that an admin can add to a domain via spec.requiredHSTSPolicies
> #### as explained in [Enforcing hsts per domain route configuration](https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-enforcing-hsts-per-domain_route-configuration) available for this
> #### Would it not be better to provide a concrete API and deprecate the annotation API?

Reviewer: ```suggestion
#### For HSTS, would it not be better to provide a concrete API and deprecate the annotation API?
```

Author: done

>   Answer: The strict-transport-security header (AKA HSTS) has wider ranging implications than the typical HTTP header. We considered allowing HSTS to be configured
>   on the IngressController in the past; see the global-options-enable-hsts EP, which was never implemented because it was superseded by
>   https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-admission-hsts.md, which was implemented.

Reviewer: ```suggestion
  Answer: The `strict-transport-security` header (for HSTS) has wider ranging implications than the typical HTTP header. We considered allowing HSTS to be configured
  on the IngressController in the past; see the global-options-enable-hsts EP, which was not accepted, and it's supercedent
  https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-admission-hsts.md, which was implemented.
```

Author: done

Reviewer 2: It seems to me that the answer is yes, we should, but doing so is beyond the scope of this EP.  Replacing the annotation with a formal API was discussed here: https://github.com/openshift/enhancements/pull/1296#discussion_r1082528269, and Joel and David both seem to favor doing so, but as Miheer said, replacing the annotation is a significant amount of work and goes beyond the requirements for this EP.

  However, HSTS isn't neatly scoped: Route A's HSTS policy can affect Route B, even if the routes are in different namespaces or belong to
  different IngressControllers, if Route B's host name is a subdomain of
  Route A's. 
  
 Additionally, specifying HSTS on an IngressController can cause a route manifest that behaves as expected on one cluster to behave surprisingly and
  problematically on another cluster. 
    

>   For these reasons, we decided not to implement the global-options-enable-hsts EP and instead wrote and implemented the
>   global-admission-hsts EP, which defines a cluster-scoped API, using which the cluster admin can specify HSTS parameters that may or must be specified on routes
>   for a given domain: https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/config/v1/types_ingress.go#L64-L87.

Reviewer: ```suggestion
  For these reasons, the global-options-enable-hsts proposal was not accepted.   The
  global-admission-hsts proposal, was accepted.  It defines a cluster-scoped API, using which the cluster admin can specify HSTS parameters that may or must be specified on routes
  for a given domain: https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/config/v1/types_ingress.go#L64-L87.
```

Author: done

  Historical note: At one time, MicroShift relied on the OpenAPI schema for server-side validation of the Route CRD, and for this reason, kubebuilder validation tags were added to the
  Route CRD (see <https://github.com/openshift/api/pull/1228>).

  Since then, an admission plugin has been added to kube-apiserver that validates the Route CRD using the same
  library-go validation on MicroShift that openshift-apiserver uses on OpenShift (see <https://github.com/openshift/kubernetes/pull/1416>).
  As a result, the kubebuilder validation tags on the Route CRD are no longer required specifically for MicroShift; however, the kubebuilder validation tags are still required for client-side validation.

#### Why don't you set/delete the headers on frontend public_ssl in the haproxy.config?
  Answer: The connection is still encrypted in public_ssl, so it is not possible to examine or modify the HTTP session there. The fe_sni and fe_no_sni frontends terminate TLS.

> #### Can we add conditions for haproxy to check to set the value? For example: Can I set the value `"%[res.hdr(server)] if { req.hdr(user-agent) -m sub evil}"` for `x-frame-options` ?

Reviewer: This shows up wrong in the md view, it is missing the word "add".  Something needs to be fixed in the formatting?

```suggestion
#### Can we use conditional values for headers?
```

>   Answer: You will be able to set the value however the if keyword won't be considered as a condition but a string by haproxy. 
>   For example: So, if you set x-frame-options: "%[res.hdr(server)] if { req.hdr(user-agent) -m sub evil}" as HTTP response then the response value served by 
>   will be `x-frame-options: Apache/2.4.37 (centos) if { req.hdr(user-agent) -m sub evil}` where we can see if condition is not evaluated as it is passed as a string to
>   haproxy.
>   
>   All values are escaped and quoted using the single quotes. Please check section `Alternative quoting strategies for haproxy.config` for more
>   details. So the value we get after escape and quote we get is a string. 

Reviewer: This is not very clear.  I think the answer is "No."
```suggestion
  Answer: No.  You cannot set a conditional value.
  For example: If you set x-frame-options: "%[res.hdr(server)] if { req.hdr(user-agent) -m sub evil}" on an HTTP x-frame-options response header, then the response value 
  will be the literal:  `x-frame-options: Apache/2.4.37 (centos) if { req.hdr(user-agent) -m sub evil}`.  In other words, the condition is not evaluated.
  
```

Author: done

### Version Skew Strategy

N/A

### Operational Aspects of API Extensions
 
#### Failure Modes
- If a user sets a header with a value which may expand to a size greater than tune.bufsize - tune.maxrewrite 
  then haproxy will send a `400` in response. For example: Setting repeatedly a dynamic value having a fetcher `ssl_c_der` and a converter `base64` 
  which exceeds the limit.
  By default, tune.maxrewrite is 8192 and tune.bufsize is 32768.
```html
<html><body><h1>400 Bad request</h1>
Your browser sent an invalid request.
</body></html>
```
Reference: https://github.com/haproxy/haproxy/issues/1309

#### Support Procedures
- If the ingress operator is in 'CrashLoopBackOff' state after setting the `httpHeaders` in the IngressController Spec then check the ingress-operator logs using `oc logs <ingress-operator-pod-name> -n openshift-ingress-operator`.
  It can go in this state if encoding of headers value was not done properly which can throw an error. You also may need to check the `ROUTER_HTTP_RESPONSE_HEADERS` and `ROUTER_HTTP_REQUEST_HEADERS`

>   vars set in the router deployment using `oc -n openshift-ingress get deployment/<router-name> -o yaml`

Reviewer: I hope this can't really happen.  The validation logic should not permit values to be submitted that will cause ingress operator or haproxy to fail to start up.

Author: It won't unless go's encode/decode function is buggy.

- Check [access logs](https://docs.openshift.com/container-platform/4.12/rest_api/operator_apis/ingresscontroller-operator-openshift-io-v1.html#spec-logging-access) of the haproxy if you want to check what HTTP Response or Request was set for the frontend and backend.

## Implementation History

## Alternatives
### Avoid Using Union Types
Using union types makes the newly added API fields more verbose and deeply nested than is strictly necessary. The API could be designed to avoid the use of unions
by changing the `IngressControllerHTTPHeader` and `RouteHTTPHeaders` types to flat structures, using CEL to require or prohibit specifying a header value based on the specified action.
In this alternative, `IngressControllerHTTPHeader` would look like the following (some extraneous details are omitted):
```go
type IngressControllerHTTPHeader struct {
    Name string `json:"name"`
    // +kubebuilder:validation:Enum:=Set;Delete
    // Action IngressControllerHTTPHeaderAction `json:"action,omitempty"`

>     // +kubebuilder:validation:XValidation:rule="has(self.value) && self.action == 'Delete'", message="Value is not required while performing delete action."

Reviewer: Nit - this doesn't line up with the updated API, which has Action rather than Value, and does use a union type.

environment variables as input parameters.

The enhancement adds two new
environment variables, `ROUTER_HTTP_RESPONSE_HEADERS` and `ROUTER_HTTP_REQUEST_HEADERS`, which specify the request and response HTTP headers to be set or deleted in the
`public`, `fe_no_sni`, and `fe_sni` frontends.

The ingress operator sets this variable based on the IngressController's
`spec.httpHeaders.actions.response[*].set` or `spec.httpHeaders.actions.response[*].delete` or `spec.httpHeaders.actions.request[*].set` or `spec.httpHeaders.actions.request[*].delete` field values. 

The value of `ROUTER_HTTP_RESPONSE_HEADERS` or `ROUTER_HTTP_REQUEST_HEADERS` is a comma-delimited list of percent encoded  triplets of the form `name:value:action` for action `Set` and `name:action` pairs for action `Delete`.
The template translates these values into the appropriate `http-response set-header` stanzas.

For example, if `ROUTER_HTTP_RESPONSE_HEADERS` has percent encoded value `"X-Frame-Options%3ADENY%3ASet%2CX-XSS-Protection%3A1%253Bmode%253Dblock%3ASet%2Cx-forwarded-client-cert%3A%2525%257B%252BQ%257D%255Bssl_c_der%252Cbase64%255D%3ASet%2CX-Frame-Options%3ADelete%2CX-XSS-Protection%3ADelete"`

> , this gets decoded into the 5 triples ("X-Frame-Options", "DENY", "Set"), ("X-XSS-Protection", "1;mode=block", "Set"), ("x-forwarded-client-cert", "%{+Q}[ssl_c_der,base64]", "Set"), ("X-Frame-Options", "Delete"), and ("X-XSS-Protection", "Delete"); and if `ROUTER_HTTP_REQUEST_HEADERS` has the percent encoded value `"Accept%3Atext%252Fplain%252C%2Btext%252Fhtml%3ASet%2CAccept-Encoding%3ADelete"`,

Reviewer 2: Wait a minute—are you _doubly_ percent-encoding these strings?!
```sh
% python3 -c 'import urllib.parse, sys; print(urllib.parse.unquote_plus(sys.argv[1]))' 'X-Frame-Options%3ADENY%3ASet%2CX-XSS-Protection%3A1%253Bmode%253Dblock%3ASet%2Cx-forwarded-client-cert%3A%2525%257B%252BQ%257D%255Bssl_c_der%252Cbase64%255D%3ASet%2CX-Frame-Options%3ADelete%2CX-XSS-Protection%3ADelete'
X-Frame-Options:DENY:Set,X-XSS-Protection:1%3Bmode%3Dblock:Set,x-forwarded-client-cert:%25%7B%2BQ%7D%5Bssl_c_der%2Cbase64%5D:Set,X-Frame-Options:Delete,X-XSS-Protection:Delete
% python3 -c 'import urllib.parse, sys; print(urllib.parse.unquote_plus(sys.argv[1]))' 'Accept%3Atext%252Fplain%252C%2Btext%252Fhtml%3ASet%2CAccept-Encoding%3ADelete'
Accept:text%2Fplain%2C+text%2Fhtml:Set,Accept-Encoding:Delete
```
Why would you _double_-encode each string?  I expected the header name and header value to be percent-encoded, but encoding the entire tuple _after_ the header name and value are already encoded seems unnecessarily complicated because then your algorithm must unquote, split, and then unquote again:
```python3
unquoted_input = urllib.parse.unquote(input)
quoted_tuples = unquoted_input.split(",")
for quoted_tuple in quoted_tuples:
    parts = quoted_tuple.split(":")
    if parts[len(parts)-1] == "Set":
        tuple = ((urllib.parse.unquote_plus(parts[0]), urllib.parse.unquote_plus(parts[1]), parts[2]))
    elif parts[len(parts)-1] == "Delete":
        tuple = ((urllib.parse.unquote_plus(parts[0]), parts[1]))
```
If you singly percent-encoded values, then the algorithm would be simpler:
```python3
tuples = input.split(",")         
for tuple in tuples:              
    parts = tuple.split(":")       
    if parts[len(parts)-1] == "Set":
        tuple = ((urllib.parse.unquote(parts[0]), urllib.parse.unquote(parts[1]), parts[2]))          
    elif parts[len(parts)-1] == "Delete":
        tuple = ((urllib.parse.unquote(parts[0]), parts[1]))     
```
And the encoded values would be easier to read—compare:
* `Accept:text%2Fplain%2C%20text%2Fhtml:Set,Accept-Encoding:Delete`
* `Accept%3Atext%252Fplain%252C%2Btext%252Fhtml%3ASet%2CAccept-Encoding%3ADelete`

Author: i made the code changes.

Author: done

   Please refer to the section `Risks and Mitigations` for more details.
   
   Only 20 items in request and response header list will be allowed.
   
   Minimum length of the header value would have to be greater than 0 and less than 16384.
   
   Allowed sample fetchers for response dynamic header values are "res.hdr" and "ssl_c_der".

   Allowed sample fetchers for request dynamic header values  are "req.hdr" and "ssl_c_der".
   
   Converters allowed for request and response are "lower" and "base64".

   Validation prohibits configuring any header actions on passthrough routes.

D) route/v1 of openshift/api will specify kubebuilder tags and CEL to enable client-side validation.

E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be the same as mentioned in point C) except the one for passthrough routes.

F) To validate the header value it has to match a regex.
For the IngressController CRD and Route CRD, this validation is defined on the request and response action lists in openshift/api.
For the Route API as handled by openshift-apiserver, this is defined in library-go.

> This will be the regex:
> ```go
> For request header:
> `^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
> ```
> ```go
> For response header:
> `^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
> ```

Reviewer 2: ````suggestion
This will be the regex
for a request header value:
```
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
And for a response header value:
```
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
````

Author: done

Author: enhancements/ingress/set-delete-http-headers.md:665 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
enhancements/ingress/set-delete-http-headers.md:669 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
enhancements/ingress/set-delete-http-headers.md:899 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]
enhancements/ingress/set-delete-http-headers.md:904 MD040/fenced-code-language Fenced code blocks should have a language specified [Context: "```"]

Author: I need to specify a language.

Reviewer 2: https://github.com/markdownlint/markdownlint/blob/main/docs/RULES.md#md040---fenced-code-blocks-should-have-a-language-specified suggests using ```` ```text ```` if there is no suitable language.  Alternatively, you can put a Go statement:
````suggestion
This will be the regex
for a request header value:
```go
permittedRequestHeaderValueRE = `^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
And for a response header value:
```go
permittedResponseHeaderValueRE = `^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
````

In this vulnerability, if the client sets the HTTP `proxy` request header and the proxy passes it through to the application, then the CGI application takes the header value and injects it in the `HTTP_PROXY` environment variable.

In this manner, an attacker could direct a vulnerable application to use a proxy of the attacker's choice.
Note that OpenShift router already mitigates this vulnerability by deleting any `proxy` header that the client specifies.

Allowing the cluster-admin or route owner to override the `proxy` HTTP header isn't necessarily a security risk; it would only allow admins or route owners to inject configuration into their own applications, which they already control.
However, the `proxy` header is non-standard, and allowing it to be configured is error-prone.

For this reason, we have not and still do not allow setting the `proxy` header

#### Strict-Transport-Security
Strict-transport-security is already [handled separately](https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-enabling-hsts_route-configuration),
so we don't allow overrides of that configuration.

> We don't allow setting the `strict-transport-security` headers on either ingress or route.

Reviewer 2: You mean IngressController, not Ingress, right?  
```suggestion
We don't allow setting the `strict-transport-security` header on either an IngressController or a Route.
```

Author: yes right

an `x-ssl` header that is controlled by the cluster-admin or route owner.  For this reason, we do not restrict setting this header.

#### Host
Setting the `host` header via the IngressController CR can cause HAProxy to fail to look up the correct route, and for that reason, we prohibit setting the `host` request header on the IngressController.

OpenShift router's `haproxy.config` file uses [the `base` sample fetch method](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#7.3.6-base) in the frontends to get the
HTTP request `host` header and path, which are then used with [the `map` converter](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#7.3.1-map) to look up the
appropriate backend to use for the request.

Modifying the `host` header on the IngressController therefore could cause HTTP requests to fail with HTTP 503 errors.
For this reason, the `host` header is in the deny list of headers that are prohibited to be set or deleted using the IngressController API.

Please note that a custom `host` header may be specified via RouteSpec. Please refer to `Open Questions` for more details.

> #### The following table summarizes as to which headers are allowed to be customized via RouteSpec or IngresscontrollerSpec
> 
> | Header name                | Existing annotation                                 | Configurable on IngressController(Using spec.httpHeaders.actions) | Configurable on Route(Using spec.httpHeaders.actions) | Configurable on IngressController(Using other API or annotation)            | Configurable on Route(Using other API or annotation) |
> |----------------------------|-----------------------------------------------------|-------------------------------------------------------------------|-------------------------------------------------------|-----------------------------------------------------------------------------|------------------------------------------------------|
> | `proxy`                    | None                                                | None                                                              | None                                                  | None                                                                        | None                                                 |
> | `host`                     | None                                                | None                                                              | Yes                                                   | None                                                                        | None                                                 |
> | `strict-transport-security` | `haproxy.router.openshift.io/hsts_header`           | None                                                              | None                                                  | Yes                                                                         | Yes                                                  |                                                     
> | `x-forwarded-for`          | `haproxy.router.openshift.io/set-forwarded-headers` | Yes                                                               | Yes                                                   | Yes(using API spec.httpHeaders.forwardedHeaderPolicy) except delete action) | Yes(using route annotation)                          |
> | `x-ssl`                    | None                                                | Yes                                                               | Yes                                                   | None                                                                        | None                                                 |                    

Reviewer 2: I think this table would make more sense under "Risks and Mitigations".

What do you think about combining the "Existing annotation", "Configurable on IngressController(Using other API or annotation)", and "Configurable on Route(Using other API or annotation)" columns?
```suggestion
The following table summarizes which headers are allowed to be customized via the Route API or IngressController API:

| Header name                 | Configurable using IngressController `spec.httpHeaders.actions` | Configurable using Route `spec.httpHeaders.actions` | Configurable using another API                                                                                                               |
|-----------------------------|-----------------------------------------------------------------|-----------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
| `proxy`                     | No                                                              | No                                                  | No                                                                                                                                           |
| `host`                      | No                                                              | Yes                                                 | No                                                                                                                                           |
| `strict-transport-security` | No                                                              | No                                                  | `haproxy.router.openshift.io/hsts_header` Route annotation                                                                                   |
| `x-forwarded-for`           | Yes                                                             | Yes                                                 | `haproxy.router.openshift.io/set-forwarded-headers` Route annotation or IngressController `spec.httpHeaders.forwardedHeaderPolicy` API field |
| `x-ssl`                     | Yes                                                             | Yes                                                 | No                                                                                                                                           |
```
| Header name                 | Configurable using IngressController `spec.httpHeaders.actions` | Configurable using Route `spec.httpHeaders.actions` | Configurable using another API                                                                                                               |
|-----------------------------|-----------------------------------------------------------------|-----------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
| `proxy`                     | No                                                              | No                                                  | No                                                                                                                                           |
| `host`                      | No                                                              | Yes                                                 | No                                                                                                                                           |
| `strict-transport-security` | No                                                              | No                                                  | `haproxy.router.openshift.io/hsts_header` Route annotation                                                                                   |
| `x-forwarded-for`           | Yes                                                             | Yes                                                 | `haproxy.router.openshift.io/set-forwarded-headers` Route annotation or IngressController `spec.httpHeaders.forwardedHeaderPolicy` API field |
| `x-ssl`                     | Yes                                                             | Yes                                                 | No                                                                                                                                           |

Author: yes i did combine this earlier. but @candita wanted me to add more columns from my understanding in slack.

Author: This looks better. i am adding this.

Reviewer: My point was that for each header that can be set in multiple places, that those places are listed.  [slack](https://redhat-internal.slack.com/archives/C04JJ0T3XT3/p1689281775921699?thread_ts=1689280690.883149&cid=C04JJ0T3XT3)

It would be even better to say which things override others, or order the columns in order of override, if there is any discernible pattern.

What @miciah listed here is still missing the HSTS API in Ingress Operator.

Author: @candita we can add it in 

Configurable using another API below haproxy.router.openshift.io/hsts_header Route annotation

Author: regarding over-ride I mentioned this in the EP in the `Open Questions` as it is  a common pattern.

### Drawbacks

Please refer to the `Risks and Mitigations` section to understand which headers are not allowed to be set/delete using this EP.

## Design Details

### Open Questions

#### Will this enhancement cover setting or deleting request headers in addition to response headers?
  Answer: Yes, we will be covering setting or deleting of request headers.
  This is required for the following use-case.  
  
  As a cluster admin or route owner, I want to enable mutual TLS but replace the "x-ssl-client-der" request header
  that OpenShift router adds with an "x-forwarded-client-cert" request header that my application checks for, 
  so that I can migrate my applications that rely on mutual TLS and use the latter header name. 

  The following example replaces "x-ssl-client-der" with "x-forwarded-client-cert":
```yaml
spec:
  httpHeaders:
    actions:

>       response:
>       - name: X-SSL-Client-Cert
>         action:
>         type: Set
>         set:
>           value: "%{+Q}[ssl_c_der,base64]"
>       - name:   Content-Language
>         action:
>         type: Delete

Reviewer 2: This example yaml is wrong.  The text above says that the _request_ headers should be modified by adding "x-forwarded-client-cert" and removing "x-ssl-client-der", so the configuration should be something more like this:
```suggestion
      request: 
      - name: x-forwarded-client-cert
        action: Set
        set:
          value: "%{+Q}[ssl_c_der,base64]"
      - name: x-ssl-client-der
        action: Delete
```

Author: done

```

#### Will this enhancement cover setting the load-balancing algorithm based on headers?
  Answer: The HAProxy configuration provides a [load balancing algorithm](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-balance) which relies on the
  check against the incoming request's header values.
  
  Even though this enhancement adds the header manipulation capability it
  doesn't aim at providing (or enhancing) the load balancing API for the IngressController or Route. 
  
  Therefore, such a load balancing algorithm is out of scope for this enhancement and should be handled with a separate RFE and thereby it will be considered as a non-goal.

#### Will this enhancement cover appending headers?

>   Answer: Yes and no.  This enhancement will not cover appending of

Reviewer 2: Why "Yes and no"?
```suggestion
  Answer: No.  This enhancement will not cover appending of
```

Author: earlier this question had "delete".   So just forgot to delete yes.

  HTTP response and request headers as per the requirements from
  [RFE-464](https://issues.redhat.com/browse/RFE-464). 
  While this RFE mentions use-cases that require the ability to set or delete headers, there is no use-case
  from this RFE that requires the ability to append headers. 

  Once such a use-case arises, we may add the option to append headers, with due
  consideration of any additional requirements deriving from the use-case
  itself or considerations around the handling of comma-separated values.  (Note
  that a use-case involving the `x-forwarded-*` headers, which is not mentioned

>   on the RFE, is described below.)

Reviewer 2: Where is this use-case mentioned?

Author: just added in For Example in the next lines

Reviewer 2: Is this what you mean by a use-case?
```markdown
  For the headers like `X-Forwarded-For` which might need values to be appended. 
  For Example:  X-Forwarded-For: <client>, <proxy1>, <proxy2>
  We have a different API for that.
```
It's difficult to understand what you are trying to say in this sub-section, or how it is different from the next sub-section ("For headers such as `X-Forwarded-For`, it makes sense to add an "Append" action. So why does this enhancement not cover appending?").  Maybe you can simplify:
```markdown
#### Will this enhancement cover appending headers?

Answer: No, this enhancement does not include an append action.
In order to keep its scope focused, the enhancement only adds set and delete actions.
These actions are sufficient to satisfy the use-cases from [RFE-464](https://issues.redhat.com/browse/RFE-464).
However, the API is designed to be extensible so that an append action, or other actions, can be added in the future, should the need arise.
```

    preserving existing headers.
    * "Replace", which specifies that the IngressController sets the headers,
    replacing any existing Forwarded or X-Forwarded-* headers.
    * "IfNone", which specifies that the IngressController sets the headers if
    they are not already set.
    * "Never", which specifies that the IngressController never sets the
    headers, preserving any existing headers. By default, the policy is
    "Append".
```

#### Are some headers blocked from being set or deleted?
  Answer: Yes, please refer to the `Risks and Mitigations` section for more details.

#### Can we please make sure to only add the `cache-control: private` header if there is no `cache-control: public` header already present in the response?
Answer:  No.  Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). 

They usually want to cache the static assets on the client and do not care about session stickiness for those requests.

Only the dynamic requests need to stick to the same server (session sticky).

Most applications will set something like this for static assets:

> ```sh
> cache-control: max-age=7200, must-revalidate, public
> ````
> 
> But the OpenShift router (HAProxy) will add cache-control: private which results in both headers in the response:
> ```sh

Reviewer 2: ````suggestion
```
cache-control: max-age=7200, must-revalidate, public
```

But OpenShift router (HAProxy) will add `cache-control: private`, which results in both headers in the response:
```
````

Author: ok

#### Should this enhancement allow setting dynamic header values?
  Answer: Yes, you would be able to set a minimal set of dynamic header values. 
  We had three options for this EP.

  1. Allow dynamic header values with all sample fetchers and converters.

  2. Allow a minimal set of dynamic values that we can confidently validate, focusing on the things that people have asked for on the RFE or [EP](https://github.com/openshift/enhancements/pull/1296#issuecomment-1396103783).

  3. Prohibit dynamic values entirely.

  However, as few customers asked for it, we went ahead with option number 2.

  For setting dynamic header values for HTTP responses, allowed sample fetchers are `res.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
  
  For setting dynamic header values for HTTP requests, allowed sample fetchers are `req.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
    
  Example: One might be able to set the value of `x-forwarded-client-cert` to `%{+Q}[ssl_c_der,base64]`.

#### For HSTS, would it not be better to provide a concrete API and deprecate the annotation API?

>   Answer:  No.  The `strict-transport-security` header (for HSTS) has wider ranging implications than the typical HTTP header.

Reviewer 2: ```suggestion
  Answer:  Yes, but HSTS requires careful treatment that goes beyond the scope of this enhancement.  
  See the discussion here: https://github.com/openshift/enhancements/pull/1296#discussion_r1082528269.
```

  details. So the value we get after escape and quote we get is a string. 

#### Why is it okay to still allow the host header on the route?
  Answer: Custom headers that are specified on the route get configured on the corresponding backend, and the backend configuration takes effect after the frontend has done its map lookup.

#### Why do we limit to max 20 items each for custom response and request headers ?
  Answer: We selected 20 items as this is a max value we can [scale](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#estimated-cost-limits) 
  w.r.t header value max size 16,384 bytes along with the CEL conditions defined.
  
  Setting a non-scalable value results in error [`Forbidden: estimated rule cost exceeds budget by factor of 2.7x (try simplifying the rule, or adding maxItems, maxProperties,
  and maxLength where arrays, maps, and strings are declared)`](https://github.com/openshift/kubernetes/blob/af29f64acbf9e6d33cde7f3194cf858f54058d3b/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1113)
  while running integration tests.

> ### Why did we select max value size of 16384 for header value ?

Reviewer: This fixes the formatting so it lines up with previous and next question.
```suggestion
#### Why did we select max value size of 16384 for header value ?
```

Author: done

this gets decoded into the two triples ("Accept", "text/plain, text/html", "Set") and ("Accept-Encoding", "Delete"); then the template adds the following stanzas to the `public`, `fe_no_sni`, and `fe_sni` frontends:
```conf
    http-response set-header X-Frame-Options 'DENY'
    http-response set-header X-XSS-Protection '1;mode=block'
    http-response set-header x-forwarded-client-cert '%{+Q}[ssl_c_der,base64]'
    http-response del-header X-Frame-Options
    http-response del-header X-XSS-Protection
    http-request set-header Accept 'text/plain, text/html'
    http-request del-header Accept-Encoding
```
Header values are quoted so that HAProxy does not misinterpret or throw an error while parsing `http-response` and `http-request` stanzas in `haproxy.config`
that contain spaces or other special characters.

Any single quotes provided within the header value will be escaped and replaced with `'\''`, and then the value is encapsulated with single quotes `''`.

> Header value  `"abc"\ "def"` is quoted as `'"abc"\ '\''def'\'''`

Author: done

> Header value  `"abc"\ "def"` is quoted as `'"abc"\ '\''def'\'''`

Reviewer 2: Fixed in https://github.com/openshift/enhancements/compare/654e8cc524e575a6009a958648baecc05f2a56e8..ecce6af5acd88f369db08634d43322c62073dc40.  Thanks!

E) operator/v1/types_ingress.go of openshift/api will contain the validations using kubebuilder tags and CEL.
The validations will be the same as mentioned in point C) except the one for passthrough routes.

F) To validate the header value it has to match a regex.
For the IngressController CRD and Route CRD, this validation is defined on the request and response action lists in openshift/api.
For the Route API as handled by openshift-apiserver, this is defined in library-go.
This will be the regex
for a request header value:
```conf
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:req\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
And for a reponse header value:
```conf
`^(?:%(?:%|(?:\{[-+]?[QXE](?:,[-+]?[QXE])*\})?\[(?:res\.hdr\([0-9A-Za-z-]+\)|ssl_c_der)(?:,(?:lower|base64))*\])|[^%[:cntrl:]])+$`
```
G) Dynamic values are allowed. (more details can be found in https://cbonte.github.io/haproxy-dconv/2.4/configuration.html#7.3) 
Example dynamic values for request:
```conf
%[req.hdr(host),lower]
%[req.hdr(host),base64]
%[req.hdr(Host)]
```
Example dynamic values for response:
```conf
%[res.hdr(X-Acl-Key),lower]
%[res.hdr(X-Acl-Key),base64]

> %[req.hdr(Host)]

Author: done

> %[req.hdr(Host)]

Reviewer 2: Fixed in https://github.com/openshift/enhancements/compare/654e8cc524e575a6009a958648baecc05f2a56e8..ecce6af5acd88f369db08634d43322c62073dc40. Thanks!

`x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
and possibly a warning on admission.

For example: If someone creates an Ingress Controller with the following:
```sh
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpHeaders:
    forwardedHeaderPolicy: Replace
```
Or if someone add a route annotation `haproxy.router.openshift.io/set-forwarded-headers`,
then the `set-header` stanzas will be added in `haproxy.config` using the [haproxy-config.template](https://github.com/openshift/router/blob/0cd9a58dff77127562f7851ccfa5cda79b8cef48/images/router/haproxy/conf/haproxy-config.template#L565-L580)

However, if someone wants to set custom values for these then they need to use the API of this EP.

> The custom values set using `spec.httpheaders.actions` will take over the value defined in annotation `haproxy.router.openshift.io/set-forwarded-headers` or `spec.httpHeaders.forwardedHeaderPolicy`.  

Reviewer 2: ```suggestion
The custom values set using `spec.httpHeaders.actions` will take over the value defined in annotation `haproxy.router.openshift.io/set-forwarded-headers` or `spec.httpHeaders.forwardedHeaderPolicy`.  
```

Author: done

  and maxLength where arrays, maps, and strings are declared)`](https://github.com/openshift/kubernetes/blob/af29f64acbf9e6d33cde7f3194cf858f54058d3b/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1113)
  while running integration tests.

#### Why did we select max value size of 16384 for header value ?
  Answer: This limit was [selected](https://www.geekersdigest.com/max-http-request-header-size-server-comparison/) as most common web servers have a limit of 16384 characters or some lower limit.
  Note that the total size of all added headers *after* interpolating dynamic values
  must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
  IngressController.

#### Why does a custom header value defined in RouteSpec override IngressControllerSpec value ?
  Answer: It's quite common to have a global setting with an escape hatch locally to override the global setting.
  The purpose of this EP is to define custom headers. So, a route owner shall have the choice to define what headers he wants as per his application requirements 

>   than the ones admin has set cluster-wide.

Reviewer 2: Out of curiosity, why did you add this open question?  In your answer, I guess you're saying there are two reasons why the Route config overrides the IngressController config: (1) The IngressController and Route might be owned by the same person who wants to specify some global configuration and then override that global configuration on a small subset of routes, and (2) Applications may have varying requirements, and the Route owner knows best what is required for the specific application.  Right?

Author: yes two reasons shall be there.

Reviewer 2: I'm still curious where this question first came up.

Reviewer: I asked about it in Slack. https://redhat-internal.slack.com/archives/C04JJ0T3XT3/p1688680813904759

#### Does a custom header value defined in RouteSpec override the route annotation ?

>   Answer: Yes. A value defined in RouteSpec overrides the route annotation.

Reviewer 2: Which route annotation are you referring to?

Reviewer 2: So you mean the custom headers specified in the Route spec override _all_ route annotations (except for `haproxy.router.openshift.io/hsts_header`, as the Route spec cannot specify `strict-transport-security`)?

* `haproxy.router.openshift.io/set-forwarded-headers`
* `haproxy.router.openshift.io/h1-adjust-case`
* `router.openshift.io/cookie_name`
* `router.openshift.io/cookie-same-site`

By the way, where this question come up?
